authz/providers/azure/rbac/rbac.go

```diff
@@ -0,0 +1,259 @@
+	// These allow us to mock out the URL for testing
+	apiURL *url.URL
+
+	tokenProvider            graph.TokenProvider
+	clusterType              string
+	azureResourceId          string
+	armCallLimit             int
+	dataStore                authz.Store
+	skipCheck                map[string]void
+	retrieveGroupMemberships bool
+	skipAuthzForNonAADUsers  bool
+}
+
+func newAccessInfo(tokenProvider graph.TokenProvider, rbacURL *url.URL, clsuterType, resourceId string, armCallLimit int, dataStore authz.Store, skipList []string, retrieveGroupMemberships, skipAuthzForNonAADUsers bool) (*AccessInfo, error) {
+	u := &AccessInfo{
+		client: http.DefaultClient,
+		headers: http.Header{
+			"Content-Type": []string{"application/json"},
+		},
+		apiURL:                   rbacURL,
+		tokenProvider:            tokenProvider,
+		azureResourceId:          resourceId,
+		armCallLimit:             armCallLimit,
+		dataStore:                dataStore,
+		retrieveGroupMemberships: retrieveGroupMemberships,
+		skipAuthzForNonAADUsers:  skipAuthzForNonAADUsers}
+
+	u.skipCheck = make(map[string]void, len(skipList))
+	var member void
+	for _, s := range skipList {
+		u.skipCheck[strings.ToLower(s)] = member
+	}
+
+	if clsuterType == "arc" {
+		u.clusterType = connectedClusters
+	}
+
+	if clsuterType == "aks" {
+		u.clusterType = managedClusters
+	}
+
+	return u, nil
+}
+
+func New(clientID, clientSecret, tenantID, aadEndpoint, armEndPoint, clusterType, resourceId string, armCallLimit int, dataStore authz.Store, skipCheck []string, retrieveGroupMemberships, skipAuthzForNonAADUsers bool) (*AccessInfo, error) {
+	rbacURL, err := url.Parse(armEndPoint)
+
+	if err != nil {
+		return nil, err
+	}
+
+	tokenProvider := graph.NewClientCredentialTokenProvider(clientID, clientSecret,
+		fmt.Sprintf("%s%s/oauth2/v2.0/token", aadEndpoint, tenantID),
+		fmt.Sprintf("%s.default", armEndPoint))
+
+	return newAccessInfo(tokenProvider, rbacURL, clusterType, resourceId, armCallLimit, dataStore, skipCheck, retrieveGroupMemberships, skipAuthzForNonAADUsers)
+}
+
+func NewWithAKS(tokenURL, tenantID, armEndPoint, clusterType, resourceId string, armCallLimit int, dataStore authz.Store, skipCheck []string, retrieveGroupMemberships, skipAuthzForNonAADUsers bool) (*AccessInfo, error) {
+	rbacURL, err := url.Parse(armEndPoint)
+
+	if err != nil {
+		return nil, err
+	}
+	tokenProvider := graph.NewAKSTokenProvider(tokenURL, tenantID)
+
+	return newAccessInfo(tokenProvider, rbacURL, clusterType, resourceId, armCallLimit, dataStore, skipCheck, retrieveGroupMemberships, skipAuthzForNonAADUsers)
+}
+
+func (a *AccessInfo) RefreshToken() error {
+	resp, err := a.tokenProvider.Acquire("")
+	if err != nil {
+		glog.Errorf("%s failed to refresh token : %s", a.tokenProvider.Name(), err.Error())
+		return errors.Wrap(err, "failed to refresh rbac token")
+	}
+
+	// Set the authorization headers for future requests
+	a.headers.Set("Authorization", fmt.Sprintf("Bearer %s", resp.Token))
```

Comment:
This call is not thread-safe as it just uses maps underneath. We need a mutex to avoid getting concurrent access panics.